hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -97,126 +280,355 @@ impl<'t> YkPTBlockIterator<'t> {
     }
 
     // Lookup a block from an offset in the "main binary" (i.e. not from a shared object).
-    fn lookup_block_from_main_bin_offset(&self, off: u64) -> Block {
+    fn lookup_block_from_main_bin_offset(&mut self, off: u64) -> Result<Block, HWTracerError> {
         if let Some(ent) = self.lookup_blockmap_entry(off) {
-            Block::from_vaddr_range(
-                u64::try_from(off_to_vaddr(&PathBuf::from(""), ent.range.start).unwrap()).unwrap(),
-                u64::try_from(off_to_vaddr(&PathBuf::from(""), ent.range.end).unwrap()).unwrap(),
-            )
+            Ok(Block::from_vaddr_range(
+                u64::try_from(self.off_to_vaddr(&PHDR_MAIN_OBJ, ent.range.start)?).unwrap(),
+                u64::try_from(self.off_to_vaddr(&PHDR_MAIN_OBJ, ent.range.end)?).unwrap(),
+            ))
         } else {
-            Block::unknown()
+            Ok(Block::unknown())
         }
     }
 
-    fn do_next(&mut self) -> Result<Option<Block>, HWTracerError> {
+    fn do_next(&mut self) -> Result<Block, HWTracerError> {
         // Read as far ahead as we can using static successor info encoded into the blockmap.
-        if let ObjLoc::MainObj(b_off) = self.cur_loc {
-            // We know where we are in the main object binary, so there's a chance that there's a
-            // blockmap entry for this location.
-            if let Some(ent) = self.lookup_blockmap_entry(b_off.to_owned()) {
-                // If there are calls in the block that come *after* the current position in the
-                // block, then we will need to follow those before we look at the successor info.
-                if let Some(call_info) = ent
-                    .value
-                    .call_offs()
-                    .iter()
-                    .find(|c| c.callsite_off() >= b_off)
-                {
-                    let target = call_info.target_off();
-                    if let Some(target_off) = target {
-                        self.cur_loc = ObjLoc::MainObj(target_off);
-                        return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
-                    } else {
-                        // Call target isn't known statically.
-                        self.cur_loc = ObjLoc::OtherObjOrUnknown;
-                        return Ok(Some(Block::unknown()));
-                    }
-                }
+        match self.cur_loc {
+            ObjLoc::MainObj(b_off) => {
+                // We know where we are in the main object binary, so there's a chance that there's a
```

Comment:
Not sure I understand this comment (is it saying "this allows to handle parts of the main executable that don't have block information"?).

Reply:
Indeed. Like the PLT code we discovered last week. I'll try a rewording at some point.

Reply:
Fixed in 53778d2bd9c41d6d5b0486c8640d65160ae55b74